lib/session.ex

```diff
@@ -0,0 +1,198 @@
+    gateway returned an error of the given code and info,
+  * other - some serious error happened.
+  """
+  @spec create_linked_session(connection_t(), timeout_t()) :: {:ok, Janus.Session.t()} | {:error, any}
+  def create_linked_session(connection, timeout \\ @default_timeout) do
+    case Connection.call(connection, %{janus: :create}, timeout) do
+      {:ok, %{"id" => session_id}} ->
+        start_link(session_id, connection)
+
+      # TODO: handle positive errors
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  @spec start_link(session_id_t(), connection_t()) :: GenServer.server()
+  def start_link(session_id, connection) do
+    GenServer.start_link(__MODULE__, {session_id, connection}, [])
+  end
+
+  @doc """
+  Synchronously attaches a plugin to the session on the gateway.
+
+  ## Arguments
+
+  * `session` - a PID of the `Janus.Session` process,
+  * `plugin` - a string containing valid gateway's plugin name,
+  * `timeout` - a timeout for the call.
+
+  ## Return values
+
+  On success it returns `{:ok, plugin_handle_id}` where `plugin_handle_id` is
+  a positive integer which is used as internal plugin handle identifier by the
+  gateway.
+
+  On error it returns `{:error, reason}`.
+
+  The reason might be:
+
+  * `{:gateway, code, info}` - it means that the call itself succeded but the
+    gateway returned an error of the given code and info,
+  * other - some serious error happened.
+  """
+  @spec session_attach(Janus.Session.t(), plugin_t(), timeout_t()) ::
+          {:ok, plugin_handle_id} | {:error, any}
+  def session_attach(session, plugin, timeout \\ @default_timeout) do
+    case execute_request(
+           session,
+           %{janus: :attach, plugin: plugin},
+           timeout
+         ) do
+      {:ok, %{"id" => plugin_handle_id}} ->
+        {:ok, plugin_handle_id}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  @doc """
+  Adds `session_id` to given message and sends it through connection stored by given session.
+
+  ## Arguments
+
+  * `session` - a PID of `Janus.Session` process.
+  * `message` - map containing message request.
+  * `timeout` - timeout passed to `Janus.Connection.call/3`.
+
+  ## Return values
+
+  Returns response same as `Janus.Connection.call/3`.
+  """
+  @spec execute_request(Janus.Session.t(), message_t(), timeout_t()) :: {:ok, any} | {:error, any}
+  def execute_request(session, message, timeout \\ @default_timeout) do
+    GenServer.call(session, {:execute_message, message, timeout})
+  end
+
+  @doc """
+  Replaces current connection with a new one.
+
+  ## Arguments
+
+  * `session` - a PID of `Janus.Session` process.
+  * `connection` - a PID of `Janus.Connection` process.
+
+  ## Return values
+
+  Returns :ok atom.
+  """
+  @spec update_connection(Janus.Session.t(), connection_t()) :: :ok
+  def update_connection(session, connection) do
+    GenServer.cast(session, {:new_connection, connection})
+  end
```

Comment:
This mechanism doesn't make much sense - Janus destroys session if connection closed, so rather than such call we need to monitor the connection and close session if the connection is down.